packages/flutter/lib/src/foundation/persistent_hash_map.dart

```diff
@@ -0,0 +1,388 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+/// A collection of key/value pairs, from which you can retrieve a value
+/// using its associated key.
+///
+/// This class implements a persistent map: extending this map with a new
+/// key/value pair does not modify an existing instance but instead creates a
+/// new instance.
+///
+/// Note: unlike [Map] this class does not support `null` as a key value and
+/// implements only a functionality needed for a specific use case at the
+/// core of the framework.
+///
+/// Underlying implementation uses a variation of *hash array mapped trie*
+/// data structure with compressed (bitmap indexed) nodes.
+///
+/// See:
+///  * [Bagwell, Phil. Ideal hash trees.](https://infoscience.epfl.ch/record/64398);
+///  * [Steindorfer, Michael J., and Jurgen J. Vinju. "Optimizing hash-array mapped tries for fast and lean immutable JVM collections."](https://dl.acm.org/doi/abs/10.1145/2814270.2814312);
+///  * [Clojure's `PersistentHashMap`](https://github.com/clojure/clojure/blob/master/src/jvm/clojure/lang/PersistentHashMap.java).
+class PersistentHashMap<K extends Object, V> {
+  /// An an empty dictionary.
+  const PersistentHashMap.empty() : this._(null);
+
+  const PersistentHashMap._(this._root);
+
+  final _TrieNode? _root;
+
+  /// Create a copy of this map extended with the given [key] to [value]
+  /// mapping.
+  PersistentHashMap<K, V> copyWith(K key, V value) {
+    final _TrieNode newRoot =
+        (_root ?? _CompressedNode.empty).copyWith(0, key, key.hashCode, value);
+    if (newRoot == _root) {
+      return this;
```

Comment:
This seems like a surprising API design that "copyWith" doesn't always copy.

Reply:
I am open to suggestions on how to name this. `updateWith` does not fit. `insert` does not fit. Maybe `extendWith` but then it does not always `extend` either, etc.

Reply:
Perhaps just `with`?

`update` would feel normal in a functional programming context.

Reply:
@lrhn suggested just to use `put`, I will rename it.